### PR TITLE
Generalize `StateTransitionEvent<S>` to allow identity transitions

### PR DESCRIPTION
**Pull request by MiniaczQ**
# Objective

This PR addresses one of the issues from [discord state discussion](https://discord.com/channels/691052431525675048/1237949214017716356).
Same-state transitions can be desirable, so there should exist a hook for them.

Fixes https://github.com/bevyengine/bevy/issues/9130.

## Solution

- Allow `StateTransitionEvent<S>` to contain identity transitions.
- Ignore identity transitions at schedule running level (`OnExit`, `OnTransition`,  `OnEnter`).
- Propagate identity transitions through `SubStates` and `ComputedStates`.
- Add example about registering custom transition schedules.

## Changelog

- `StateTransitionEvent<S>` can be emitted with same `exited` and `entered` state.

### CODE REVIEW

**Comment by MiniaczQ** [comment]
Just waiting on #13626, then this can proceed and I'll need to draft one more for `ComputedStates` extension

**Comment by MiniaczQ** [comment]
I just need to add a test for `ComputedStates` propagation, but the bulk of the work is done and ready for review.

File: examples/state/custom_transitions.rs [hunk 1]
```diff
@@ -0,0 +1,273 @@
+//! This example illustrates how to register custom state transition behavior.
+//!
+//! In this case we are trying to add `OnReenter` and `OnReexit`
+//! which will work much like `OnEnter` and `OnExit`,
+//! but additionally trigger if the state changed into itself.
+//!
+//! While identity transitions exist internally in [`StateTransitionEvent`]s,
+//! the default schedules intentionally ignore them, as it is not the common use-case.
```
Comment (alice-i-cecile) [inline examples/state/custom_transitions.rs:8]:
```suggestion
//! the default schedules intentionally ignore them, as this behavior is not commonly needed or expected.
```

File: examples/state/custom_transitions.rs [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        };
+
+        sprite.color = new_color.into();
+    }
+}
+
+// We can restart the game by pressing "R".
+// This will trigger an [`AppState::InGame`] -> [`AppState::InGame`]
+// transition, which will run our custom schedules.
+fn restart_game(input: Res<ButtonInput<KeyCode>>, mut next_state: ResMut<NextState<AppState>>) {
```
Comment (alice-i-cecile) [inline examples/state/custom_transitions.rs:116]: I think this system's name will be much clearer as `trigger_game_restart`.

File: examples/state/custom_transitions.rs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        let _ = world.try_run_schedule(OnReexit(exited));
+    }
+
+    /// The plugin registers the transitions for one specific state.
+    /// If you use this for multiple states consider:
+    /// - installing the plugin multiple times,
+    /// - create an [`App`] extension method that inserts
+    ///   those transitions during state installation.
+    #[derive(Default)]
+    pub struct IdentityTransitionsPlugin<S: States>(PhantomData<S>);
```
Comment (alice-i-cecile) [inline examples/state/custom_transitions.rs:178]: I think the flow will be clearer with this definition at the start of this submodule.

File: crates/bevy_state/src/state/transitions.rs [hunk 4]
```diff
@@ -88,14 +93,15 @@ pub(crate) fn internal_apply_state_transition<S: States>(
...
-                            exited: Some(exited.clone()),
-                            entered: Some(entered.clone()),
-                        });
-                    }
+                    let exited = match *state_resource == entered {
+                        true => entered.clone(),
+                        false => mem::replace(&mut state_resource.0, entered.clone()),
+                    };
+
+                    event.send(StateTransitionEvent {
```
Comment (alice-i-cecile) [inline crates/bevy_state/src/state/transitions.rs:101]:
```suggestion
                    // Transition events are sent even for same state transitions
                    // Although enter and exit schedules are not run by default.
                    event.send(StateTransitionEvent {
```

**Review by lee-orr — APPROVED** [review]
Agreed about liking this approach! It addresses alot of my concerns around same-state transitions while still enabling those use cases where it's relevant.

File: examples/state/custom_transitions.rs [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+// We can restart the game by pressing "R".
+// This will trigger an [`AppState::InGame`] -> [`AppState::InGame`]
+// transition, which will run our custom schedules.
+fn trigger_game_restart(
+    input: Res<ButtonInput<KeyCode>>,
+    mut next_state: ResMut<NextState<AppState>>,
+) {
+    if input.just_pressed(KeyCode::KeyR) {
+        // Although we are already in this state setting it again will generate an identity transition.
+        // While default schedules ignore those kinds of transitions, out custom schedules will react to them.
```
Comment (alice-i-cecile) [inline examples/state/custom_transitions.rs:220]:
```suggestion
        // While default schedules ignore those kinds of transitions, our custom schedules will react to them.
```

File: examples/state/custom_transitions.rs [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+                StateTransition,
+                // The internals can generate at most one transition event of specific type per frame.
+                // We take the latest one and clear the queue.
+                last_transition::<S>
+                    // We insert the optional event into our schedule runner.
+                    .pipe(run_reenter::<S>)
+                    // We need to pick at which step the schedule will be executed.
+                    // There are 3 intended steps for custom schedules.
+                    // They are executed in that order and built-in schedules
+                    // belong to them respectively. Those sets are:
```
Comment (alice-i-cecile) [inline examples/state/custom_transitions.rs:75]:
```suggestion
                    // State transitions are handled in three ordered steps, exposed as system sets.
                    // We can add our systems to them, which will run the corresponding schedules when they're evaluated.
                    // These are:
```

**Review by alice-i-cecile — APPROVED** [review]
One more wording suggestion, but I'm happy with this!

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1348 if you'd like to help out.

**Pull request merged**